pkg/storkctl/clusterpair.go

```diff
@@ -476,51 +752,124 @@ func getConfig(configFile string) clientcmd.ClientConfig {
 }
 
-func getClusterPairParams(config, endpoint string, customPort string) (string, string, string, error) {
-	var ip, port, token string
+func getPXEndPointDetails(config string) (string, error) {
+	var ep string
 	client, err := core.NewInstanceFromConfigFile(config)
 	if err != nil {
-		return ip, port, token, err
+		return ep, err
 	}
 
 	services, err := client.ListServices("", meta.ListOptions{LabelSelector: "name=portworx-api"})
 	if err != nil || len(services.Items) == 0 {
-		err := fmt.Errorf("unable to retrieve portworx-api service from DR cluster. Err: %v", err)
-		return ip, port, token, err
+		err := fmt.Errorf("unable to retrieve portworx-api service. Err: %v", err)
+		return ep, err
 	}
-	// TODO: in case of setting up aync-dr over cloud,
-	// users set up different service as load-balancer over px apis
-	// accept px-service name as env variable
+
+	// Currently we are handling only the cases where portworx-api service is either of type LoadBalancer or clusterIP.
+	// For custom upstream network objects like route and ingress, User has to provide the endpoints.
 	svc := services.Items[0]
-	ip = endpoint
-	if ip == "" {
-		// this works only if px service is converted as load balancer type
-		// TODO: for 2 cluster where worker nodes are reachable, figure out
-		// any one worker ip by looking at px/enabled label
-		ip = svc.Spec.LoadBalancerIP
-	}
-	pxToken := os.Getenv("PX_AUTH_TOKEN")
-	if len(customPort) > 0 {
-		port = customPort
-	} else {
+	// Only if px svc is load balancer type get the IP or Host details
+	if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+		if len(svc.Status.LoadBalancer.Ingress) > 0 {
+			if len(svc.Status.LoadBalancer.Ingress[0].Hostname) > 0 {
+				ep = svc.Status.LoadBalancer.Ingress[0].Hostname
+			} else if len(svc.Status.LoadBalancer.Ingress[0].IP) > 0 {
+				ep = svc.Status.LoadBalancer.Ingress[0].IP
+			}
+		}
+
 		for _, svcPort := range svc.Spec.Ports {
 			if svcPort.Name == "px-api" {
-				port = strconv.Itoa(int(svcPort.Port))
+				port := strconv.Itoa(int(svcPort.Port))
+				ep = net.JoinHostPort(ep, port)
 				break
 			}
 		}
+	} else if svc.Spec.Type == v1.ServiceTypeClusterIP {
+		endPoints, err := client.GetEndpoints("portworx-api", svc.Namespace)
+		if err != nil {
+			return ep, fmt.Errorf("unable to retrieve portworx-api endpoint. Err: %v", err)
+		}
+		if len(endPoints.Subsets) > 0 && len(endPoints.Subsets[0].Addresses) > 0 {
+			ip := endPoints.Subsets[0].Addresses[0].IP
+			for _, epPort := range endPoints.Subsets[0].Ports {
+				if epPort.Name == "px-api" {
+					port := strconv.Itoa(int(epPort.Port))
+					ep = net.JoinHostPort(ip, port)
+				}
+			}
+		}
+	} else {
+		return ep, fmt.Errorf("an explicit portworx endpoint is required when portworx-api service type is %s", svc.Spec.Type)
 	}
-	pxEndpoint := net.JoinHostPort(ip, port)
+	return ep, nil
+}
+
+func getPXToken(pxEndpoint, pxToken string) (string, error) {
+	var token string
+
+	net.SplitHostPort(pxEndpoint)
```

Comment:
Can you explain the change in this line?

Reply:
This is not required. I forgot to remove this after moving the `SplitHostPort ` to another function `getHostPortFromEndPoint`